### PR TITLE
Added kwarg support for dynamo.compile

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Please review my changes for adding kwarg support for dynamo.compile

### CODE REVIEW

File: tests/py/dynamo/models/test_models_export_kwargs.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        "optimization_level": 1,
+        "min_block_size": 1,
+        "ir": "dynamo",
+    }
+    # TODO: Support torchtrt.compile
+    # trt_mod = torchtrt.compile(model, **compile_spec)
+
+    exp_program = torch.export.export(model, args=tuple(args), kwargs=kwargs)
+    trt_mod = torchtrt.dynamo.compile(exp_program, **compile_spec)
+    cos_sim = cosine_similarity(model(*args, **kwargs), trt_mod(*args, **kwargs)[0])
```
Comment (peri044) [inline tests/py/dynamo/models/test_models_export_kwargs.py:66]: trt_mod(*args, **kwargs)[0] - `[0]` is needed here ?

File: tests/py/dynamo/models/test_models_export_kwargs.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    assertions.assertTrue(
+        cos_sim > COSINE_THRESHOLD,
+        msg=f"Resnet18 TRT outputs don't match with the original model. Cosine sim score: {cos_sim} Threshold: {COSINE_THRESHOLD}",
```
Comment (peri044) [inline tests/py/dynamo/models/test_models_export_kwargs.py:69]: Rename `Resnet18`to maybe `CustomKwargs` or something

File: py/torch_tensorrt/dynamo/utils.py [hunk 3]
```diff
@@ -128,24 +128,45 @@ def input_is_dynamic(inputs: Sequence[Union[Input, torch.Tensor]]) -> bool:
 
 
 def get_torch_inputs(
-    inputs: Sequence[Input], device: Union[Device, torch.device, str], mode: str = ""
-) -> Sequence[torch.tensor]:
+    inputs: Sequence[Input] | Dict[Any, Any],
+    device: Union[Device, torch.device, str],
+    mode: str = "",
```
[the remark this answers is not in the document]
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/utils.py:133]: That is there before my changes. I am not quite sure. Should I remove it?

File: tests/py/dynamo/models/test_models_export_kwargs.py (hunk 1 above)
Comment (cehongwang) [reply tests/py/dynamo/models/test_models_export_kwargs.py:66]: I think trt_module output as a list. [0] gets the output tensor itself

**Comment by narendasan** [comment]
There are other APIs like `torch_tensorrt.save` that use a similar input arg. we should update those to match

File: py/torch_tensorrt/_compile.py [hunk 4]
```diff
@@ -408,6 +408,7 @@ def save(
     *,
     output_format: str = "exported_program",
     inputs: Optional[Sequence[torch.Tensor]] = None,
```
Comment (narendasan) [inline py/torch_tensorrt/_compile.py:410]: Should add the arg_inputs argument + handling here as well

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 5]
```diff
@@ -47,8 +48,9 @@
 
 def compile(
     exported_program: ExportedProgram,
-    inputs: Tuple[Any, ...],
+    inputs: Sequence[Any],
     *,
+    kwarg_inputs: Optional[dict[Any, Any]] = None,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:53]: Can we add an arg_inputs parameter?

File: docs/_downloads/7e3a125a2d4ba8274a41b46f5e0723fa/refit_engine_example.py [hunk 6]
```diff
@@ -78,7 +78,7 @@
 new_trt_gm = refit_module_weights(
```
Comment (narendasan) [inline docs/_downloads/7e3a125a2d4ba8274a41b46f5e0723fa/refit_engine_example.py:78]: dont commit changes in docs/

**Review by narendasan — COMMENTED** [review]
Seems fine to me, @peri044, @zewenli98 can you take a look?

File: py/torch_tensorrt/_compile.py [hunk 7]
```diff
@@ -238,17 +242,33 @@ def compile(
         return compiled_fx_module
     elif target_ir == _IRType.dynamo:
         # Prepare torch and torchtrt inputs
+        if not arg_inputs and not inputs:
+            raise AssertionError("'arg_input' and 'input' should not both be None.")
+
+        elif arg_inputs and inputs:
+            raise AssertionError(
+                "'arg_input' and 'input' should not be used at the same time."
```
Comment (peri044) [inline py/torch_tensorrt/_compile.py:250]: minor comment: `arg_inputs` and `inputs`

File: py/torch_tensorrt/_compile.py [hunk 8]
```diff
@@ -342,18 +366,35 @@ def convert_method_to_trt_engine(
         )
     elif target_ir == _IRType.dynamo:
         # Prepare torch and torchtrt inputs
+        if not arg_inputs and not inputs:
+            raise AssertionError("'arg_input' and 'input' should not both be None.")
+
+        elif arg_inputs and inputs:
+            raise AssertionError(
+                "'arg_input' and 'input' should not be used at the same time."
+            )
```
Comment (peri044) [inline py/torch_tensorrt/_compile.py:375]: minor comment: `arg_inputs` and `inputs`

File: py/torch_tensorrt/_compile.py [hunk 9]
```diff
@@ -416,18 +459,32 @@ def save(
...
+    if arg_inputs is not None and not all(
+        isinstance(input, torch.Tensor) for input in arg_inputs
     ):
         raise ValueError(
             "Not all inputs provided are torch.tensors. Please provide torch.tensors as inputs"
         )
+    if arg_inputs and inputs:
+        raise AssertionError(
+            "'arg_input' and 'input' should not be used at the same time."
+        )
```
Comment (peri044) [inline py/torch_tensorrt/_compile.py:479]: same as above

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 10]
```diff
@@ -181,27 +185,41 @@ def compile(
...
 
-    if not isinstance(inputs, collections.abc.Sequence):
-        inputs = [inputs]
+    # Aliasing inputs to arg_inputs for better understanding
+    if not arg_inputs and not inputs:
+        raise AssertionError("'arg_input' and 'input' should not both be None.")
+
+    elif arg_inputs and inputs:
+        raise AssertionError(
+            "'arg_input' and 'input' should not be used at the same time."
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_compiler.py:194]: same as above

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 11]
```diff
@@ -579,13 +608,25 @@ def convert_module_to_trt_engine(
             DeprecationWarning,
             stacklevel=2,
         )
+    if not arg_inputs and not inputs:
+        raise AssertionError("'arg_input' and 'input' should not both be None.")
+
+    elif arg_inputs and inputs:
+        raise AssertionError(
+            "'arg_input' and 'input' should not be used at the same time."
+        )
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_compiler.py:617]: same as above

File: py/torch_tensorrt/dynamo/_tracer.py [hunk 12]
```diff
@@ -52,20 +58,77 @@ def trace(
     """
 
     # Set log level at the top of compilation (torch_tensorrt.dynamo)
+    if not arg_inputs and not inputs:
+        raise AssertionError("'arg_input' and 'input' should not both be None.")
+
+    elif arg_inputs and inputs:
+        raise AssertionError(
+            "'arg_input' and 'input' should not be used at the same time."
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_tracer.py:66]: same as mentioned above

File: py/torch_tensorrt/dynamo/_tracer.py [hunk 13]
```diff
@@ -52,20 +58,77 @@ def trace(
...
     device = to_torch_device(kwargs.get("device", default_device()))
-    torch_inputs = get_torch_inputs(inputs, device)
-    dynamic_shapes = []
-    for input in inputs:
-        if isinstance(input, Input) and input.shape_mode == Input._ShapeMode.DYNAMIC:
+    torch_arg_inputs = get_torch_inputs(arg_inputs, device)
+    torch_kwarg_inputs = get_torch_inputs(kwarg_inputs, device)
+    dynamic_shapes = get_dynamic_shapes_args(mod, arg_inputs)
+    dynamic_shapes.update(get_dynamic_shapes_kwargs(kwarg_inputs))
+    # breakpoint()
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/_tracer.py:82]: remove this comment

File: py/torch_tensorrt/dynamo/conversion/_conversion.py [hunk 14]
```diff
@@ -27,12 +27,16 @@ def infer_module_output_dtypes(
...
     inputs: Sequence[Input],
     device: Device,
+    kwarg_inputs: Optional[dict[str, Any]] = None,
     truncate_double: bool = False,
 ) -> List[dtype]:
     with maybe_disable_fake_tensor_mode():
         torch_inputs = get_torch_inputs(inputs, device)
+        if kwarg_inputs is None:
+            kwarg_inputs = {}
+        torch_kwarg_inputs = get_torch_inputs(kwarg_inputs, device)
```
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/_conversion.py:37]: Consider handling the None case within `get_torch_inputs` call itself.

File: tests/py/dynamo/models/test_export_serde.py [hunk 15]
```diff
@@ -13,232 +13,272 @@
...
-    #         cos_sim > COSINE_THRESHOLD,
-    #         msg=f"test_base_full_compile_multiple_outputs deserialized TRT outputs don't match with the original model. Cosine sim score: {cos_sim} Threshold: {COSINE_THRESHOLD}",
-    #     )
+# @pytest.mark.unit
+# def test_base_full_compile(ir):
+#     """
+#     This tests export serde functionality on a base model
+#     which is fully TRT convertible
+#     """
+
```
Comment (peri044) [inline tests/py/dynamo/models/test_export_serde.py:22]: why are all these commented out ?

**Comment by cehongwang** [comment]
@narendasan @peri044 @zewenli98 If there is no further changes required, should I merge this now?

File: py/torch_tensorrt/dynamo/utils.py [hunk 16]
```diff
@@ -387,3 +411,31 @@ def check_output(
                 return False
 
     return True
+
+
+def flatten_dict_value(d: dict[Any, Any]) -> List[Any]:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/utils.py:416]: We need to check the determinism here.

File: py/torch_tensorrt/dynamo/utils.py [hunk 17]
```diff
@@ -128,24 +128,45 @@ def input_is_dynamic(inputs: Sequence[Union[Input, torch.Tensor]]) -> bool:
...
     """
     Return the torch_tensor from the Input object. If mode is set, this implies
     user is using dynamic shaped inputs and return the corresponding input based
     on the mode requested.
     """
     device = to_torch_device(device)
     if mode:
+        if isinstance(inputs, dict):
+            result = {}
+            for k, v in inputs.items():
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/utils.py:144]: Check the need for determinism here

File: py/torch_tensorrt/dynamo/_tracer.py [hunk 18]
```diff
@@ -52,20 +58,78 @@ def trace(
...
+        dynamic_shapes=dynamic_shapes,
+    )
+
+    return exp_program
+
+
+def get_dynamic_shapes_kwargs(inputs: Any) -> Union[dict[str, Any], list[Any]]:
+    if isinstance(inputs, dict):
+        dynamic_shapes_kwarg = {}
+        for k, v in inputs.items():
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_tracer.py:96]: Check need for determinism here:

**Review by narendasan — COMMENTED** [review]
Make sure that the torch.export kwargs ordering is matched by our flattening or use their flattening

**Review by narendasan — APPROVED** [review]
LGTM

**Pull request merged**